pilot/pkg/serviceregistry/kube/controller/filter/discoverynamespaces.go

```diff
@@ -0,0 +1,65 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package filter
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/sets"
+	v1 "k8s.io/client-go/listers/core/v1"
+
+	"istio.io/pkg/log"
+)
+
+const (
+	// PilotDiscoveryLabelName is the name of the label used to indicate a namespace for discovery
+	PilotDiscoveryLabelName = "istio-discovery"
+	// PilotDiscoveryLabelValue is the value of for the label PilotDiscoveryLabelName used to indicate a namespace for discovery
+	PilotDiscoveryLabelValue = "true"
+)
+
+var PilotDiscoverySelector = labels.Set(map[string]string{PilotDiscoveryLabelName: PilotDiscoveryLabelValue}).AsSelector()
+
+type Func func(obj interface{}) bool
+
+var permitAll = func(_ interface{}) bool {
+	return true
+}
+
+func DiscoveryNamespacesFilterFactory(lister v1.NamespaceLister, enableDiscoveryNamespaces bool) func() Func {
```

Comment:
why so complicated func() Func

Reply:
This was in response to [John's comment above](https://github.com/istio/istio/pull/29802#discussion_r549812075). The factory pattern is an optimization that allows us to initialize the filter (and iterate the namespaces) [once per invocation](https://github.com/istio/istio/pull/29802/files#diff-9b1e89a7f445a55fef998ca65bf562bab5f072b96af6b5326efce292eec2f72bR100-R101) to `List()`.

I agree it's complicated, I'm open to other approaches as well

Reply:
This has been simplified under the new cache approach (commit `a3d6016`)